### PR TITLE
Add organization account e2e test

### PR DESCRIPTION
**Pull request by adrien-layer**
We add first organization e2e test for:
- building a dataset
- training a model

(we reuse existing test methods to improve maintainability and parity checks with personal accounts)

- We create only one organization account but one project for each test.
- The test creates and cleans up the organization account.

### CODE REVIEW

**Comment by adrien-layer** [comment]
failing because we need a new token with new org account claim

need to relogin / refresh token after org account creation

**Review by marta-tatiana — APPROVED** [review]

File: test/e2e/test_organization_function_run.py
Comment (dpp23) [inline test/e2e/test_organization_function_run.py:42]: Is this going to cause races between different tests trying to refresh the login?

For guest logins, we had to do a per-test config.json. see https://github.com/layerai/sdk/blob/main/test/e2e/conftest.py#L148

File: test/e2e/test_organization_function_run.py
Comment (dpp23) [inline test/e2e/test_organization_function_run.py:38]: Are all users allowed to create an organization account?

File: test/e2e/test_organization_function_run.py
Comment (adrien-layer) [reply test/e2e/test_organization_function_run.py:42]: yes i wanted to test before going down that route, but it is likely we'll need to

File: test/e2e/test_organization_function_run.py
Comment (adrien-layer) [reply test/e2e/test_organization_function_run.py:38]: atm yes, it will evolve later when we add restrictions

Comment (dpp23) [reply test/e2e/test_organization_function_run.py:38]: Not a blocker but one thing to keep in mind so we update the docs when we get there - the current assumption is that anyone can run e2e-tests with their personal account. If the user needs special permissions, then we will need to make it explicit that `e2e-tests` are only for internal use.

We can just add a flag/selector to skip them as well.

Comment (adrien-layer) [reply test/e2e/test_organization_function_run.py:38]: do you think it is worth adding that flag now to skip them when choosing to?

im thinking when time comes it will be part of the task to do that

File: test/e2e/test_organization_function_run.py
Comment (adrien-layer) [reply test/e2e/test_organization_function_run.py:42]: actually come to think of it: the permissions needed there are additive, the only risk i see is that we invalidate the token of other tests but other tests actually do a config refresh (without force) as part of their fixture chaining, so i think this should be okay

File: test/e2e/test_organization_function_run.py
Comment (dpp23) [reply test/e2e/test_organization_function_run.py:35]: No, let's leave it for later when it is actually a problem. Just wanted to mention it for future consideration, no need to do this now.

File: test/e2e/test_organization_function_run.py
Comment (adrien-layer) [reply test/e2e/test_organization_function_run.py]: this is irrelevant now that we use the pytest_sessionstart hook

File: test/e2e/conftest.py
Comment (dpp23) [inline test/e2e/conftest.py:146]: Are we sure this will get executed at least once?

File: test/e2e/conftest.py
Comment (dpp23) [inline test/e2e/conftest.py:52]: I don't quite follow what is the sequence here and what we are trying to solve by this file signaling.

Is `pytest_sessionfinish` called on the main process before all tests are finished?

File: test/e2e/conftest.py
Comment (adrien-layer) [reply test/e2e/conftest.py:146]: it should yes since the flag is set by `pytest_sessionfinish` with '`a`' append mode

what could happen though is that `pytest_sessionfinish` suddenly starts being executed after fixture teardowns, so i can add an attempt to delete the org account also inside the hook

File: test/e2e/conftest.py
Comment (adrien-layer) [reply test/e2e/conftest.py:52]: yes kinda, meaning fixture teardowns are executed after `pytest_sessionfinish` . i should improve the explanation here then

File: test/e2e/conftest.py
Comment (dpp23) [reply test/e2e/conftest.py:146]: Are fixtures cleaned up after every test or at the end of the run?

What will happen if the tests get split like this:

w0 - orgTest, regularTest
w1 - orgTest, regularTest

?

If the tear down happens after `orgTest`, that will be before `pytest_sessionfinish` so we will never clean up.

Comment (adrien-layer) [reply test/e2e/conftest.py:146]: i dont understand the question sorry. currently there is an autoUse fixture that will be run after every test, so one single test teardown will be the one to execute the session cleanup

Comment (dpp23) [reply test/e2e/conftest.py:146]: oh, I see, so that cleanup is attempted after every test regardless of whether it actually uses the org or not.

Comment (adrien-layer) [reply test/e2e/conftest.py:146]: yes indeed, this is why I call it `_test_session_teardown`, kinda like our own hook for our distributed use case.

i thought `pytest_sessionfinish` was that but it isnt behaving as i expected

File: test/e2e/conftest.py
Comment (adrien-layer) [reply test/e2e/conftest.py]: found this https://github.com/pytest-dev/pytest/issues/7484, last commit uses it

File: test/e2e/conftest.py
Comment (adrien-layer) [reply test/e2e/conftest.py]: using this https://github.com/pytest-dev/pytest/issues/7484 now

**Review by dpp23 — APPROVED** [review]

**Comment by adrien-layer** [comment]
org accounts are actually not cleaned up

seems pytest_sessionfinish simply does not run with xdist

im gonna use a file lock now because these hooks are too unpredictable
also this is what xdist suggests https://pypi.org/project/pytest-xdist/#making-session-scoped-fixtures-execute-only-once

**Comment by adrien-layer** [comment]
hooray, now organization account is being cleaned up properly. also tested locally.

after merge i will monitor this is reliable

File: test/e2e/conftest.py
Comment (adrien-layer) [inline test/e2e/conftest.py:59]: fix typo

Comment (dpp23) [inline test/e2e/conftest.py]:
```suggestion
    we need extra logic to do track parallel pytest sessions and do teardown once there are no more.
```

File: test/e2e/conftest.py
Comment (adrien-layer) [inline test/e2e/conftest.py:166]: do this first so we avoid setting up the client for nothing

File: test/e2e/conftest.py
Comment (dpp23) [inline test/e2e/conftest.py:47]: If I get in this scenario, how do I unblock myself? Would it be better if we just try to delete it and move on? Maybe error if we fail?

File: test/e2e/conftest.py
Comment (adrien-layer) [reply test/e2e/conftest.py:64]: i thought about this and thought we'd still need to cleanup but we do want to be aware yes, will add sthg

File: test/e2e/conftest.py
Comment (adrien-layer) [reply test/e2e/conftest.py:47]: this was to prevent a bad setup and should fail early. but i should delete then before failing so we don't leak.
i would still have this fail early instead of silently ignore this illegal state

**Pull request merged**
